osquery/tables/ipmi/ipmi_info.cpp

```diff
@@ -0,0 +1,1309 @@
+          d->client->insertRowsQueue(
+              d->mcKey, d->colName, toSetInProgress(parmVal, dataLen));
+          break;
+
+        default:
+          LOG(ERROR) << "Got an unexpected LANPARM type: "
+                     << std::to_string(d->parm);
+          break;
+        }
+
+      },
+      data);
+
+  if (rv != 0) {
+    LOG(ERROR) << "Could not get default gateway lanparm: " << strerror(rv);
+  }
+}
+
+/*===========================OpenIPMI Callbacks================================
+ */
+
+void getLANsCB(ipmi_domain_t* domain, ipmi_mc_t* mc, void* data) {
+  IPMIClient* client = (IPMIClient*)data;
+
+  std::string key = getMCKey(mc);
+  ipmi_lanparm_t* lp = client->getLANParm(key, mc);
+  if (lp == nullptr) {
+    LOG(ERROR) << "Could not get ipmi_lanparm_t for mc '" << key;
+    return;
+  }
+
+  char name[IPMI_MC_NAME_LEN];
+  int len = ipmi_mc_get_name(mc, name, IPMI_MC_NAME_LEN);
+  if (len > 0) {
+    client->insertRowsQueue(key, "mc_name", name);
+  }
+  client->insertRowsQueue(key, "mc_id", std::to_string(ipmi_mc_device_id(mc)));
+
+  getPARM(lp, IPMI_LANPARM_IP_ADDRESS, key, "ip", client);
+
+  getPARM(lp, IPMI_LANPARM_DEFAULT_GATEWAY_ADDR, key, "gateway_ip", client);
+
+  getPARM(lp, IPMI_LANPARM_MAC_ADDRESS, key, "mac", client);
+
+  getPARM(lp, IPMI_LANPARM_SUBNET_MASK, key, "subnet_mask", client);
+
+  getPARM(
+      lp, IPMI_LANPARM_DEFAULT_GATEWAY_MAC_ADDR, key, "gateway_mac", client);
+
+  getPARM(
+      lp, IPMI_LANPARM_BACKUP_GATEWAY_ADDR, key, "backup_gateway_ip", client);
+
+  getPARM(lp,
+          IPMI_LANPARM_BACKUP_GATEWAY_MAC_ADDR,
+          key,
+          "backup_gateway_mac",
+          client);
+
+  getPARM(lp, IPMI_LANPARM_IP_ADDRESS_SRC, key, "ip_address_source", client);
+
+  getPARM(
+      lp, IPMI_LANPARM_COMMUNITY_STRING, key, "snmp_community_string", client);
+
+  getPARM(lp, IPMI_LANPARM_SET_IN_PROGRESS, key, "set_in_progress", client);
+}
+
+void IPMIFullyUpCB(ipmi_domain_t* domain, void* data) {
+  TLOG << "OpenIPMI is now fully up";
+  IPMIClient* c = (IPMIClient*)data;
+  c->setDomain(domain);
+
+  return;
+}
+
+/// Gets the the value suffix for a sensor reading.
+std::string getSensorThresholdSuffix(ipmi_sensor_t* sensor) {
+  const size_t suffixLen = 50;
+  char suffix[suffixLen];
+
+  const char* percent = "";
+  const char* base;
+  const char* mod_use = "";
+  const char* modifier = "";
+  const char* rate;
+  base = ipmi_sensor_get_base_unit_string(sensor);
+  if (ipmi_sensor_get_percentage(sensor))
+    percent = "%";
+  switch (ipmi_sensor_get_modifier_unit_use(sensor)) {
+  case IPMI_MODIFIER_UNIT_NONE:
+    break;
+  case IPMI_MODIFIER_UNIT_BASE_DIV_MOD:
+    mod_use = "/";
+    modifier = ipmi_sensor_get_modifier_unit_string(sensor);
+    break;
+  case IPMI_MODIFIER_UNIT_BASE_MULT_MOD:
+    mod_use = "*";
+    modifier = ipmi_sensor_get_modifier_unit_string(sensor);
+    break;
+  }
+  rate = ipmi_sensor_get_rate_unit_string(sensor);
+
+  snprintf(
```

Comment:
I can't help but feel that you could restructure the logic of this function to make use of modern C++ data structures, like `std::vector`s of `std::string`s, along side `osquery::join` or the `std::string` `+` operator? Seems to me like there's a lot of coercion to get C-style data structures working here.